presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunction.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.storage.organization;
+
+import com.facebook.presto.raptor.storage.StorageManagerConfig;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import org.joda.time.DateTimeZone;
+
+import javax.inject.Inject;
+
+import java.time.Duration;
+
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TemporalFunction
+{
+    private final DateTimeZone timeZone;
+
+    @Inject
+    public TemporalFunction(StorageManagerConfig config)
+    {
+        this(config.getShardDayBoundaryTimeZone());
+    }
+
+    public TemporalFunction(DateTimeZone timeZone)
+    {
+        this.timeZone = requireNonNull(timeZone, "timeZone is null");
+    }
+
+    public int getDay(Type type, Block block, int position)
+    {
+        if (type.equals(DATE)) {
+            return toIntExact(DATE.getLong(block, position));
+        }
+
+        if (type.equals(TIMESTAMP)) {
+            long millis = block.getLong(position, 0);
```

Comment:
This is wrong. You should never read from blocks directly. Always use `Type` to access the block, as types own the data representation within a block.